syntax = "proto3";

// GetLabyrinthInfo, GetPlayerStatus, RegisterMove are simple unary RPCs
// Revelio is a response-streaming RPC
// Bombarda is a request-streaming RPC

service LabyrinthService {
    rpc GetLabyrinthInfo(Empty) returns (LabyrinthInfo);
    rpc GetPlayerStatus(Empty) returns (PlayerStatus);
    rpc RegisterMove(MoveRequest) returns (MoveResponse);
    rpc Revelio(RevelioRequest) returns (stream TilePosition);
    rpc Bombarda(stream TargetPosition) returns (BombardaResponse);
    rpc GetGrid(Empty) returns (Grid);
}

message Empty {}

message LabyrinthInfo {
    int32 width = 1;
    int32 height = 2;
}

message PlayerStatus {
    int32 x = 1;
    int32 y = 2;
    int32 hp = 3;
    int32 rem_spells = 4;
    int32 score = 5;
}

message MoveRequest {
    string direction = 1;
}

message MoveResponse {
    int32 status = 1;
}

message TilePosition {
    int32 x = 1;
    int32 y = 2;
}

message TargetPosition {
    int32 x = 1;
    int32 y = 2;
}

message RevelioRequest {
    int32 x = 1;
    int32 y = 2;
    int32 tile_type = 3;
}

message BombardaResponse {
    string status = 1;
}

message Grid {
    repeated GridRow gridrow = 1;
}

message GridRow {
    repeated string val = 1;
}
